internal/querynodev2/segments/search.go

```diff
@@ -109,6 +109,78 @@ func searchSegments(ctx context.Context, mgr *Manager, segments []Segment, segTy
 	return searchResults, nil
 }
 
+const (
+	defaultStreamConcurrency = 2
+)
+
+// searchSegmentsStreamly performs search on listed segments in a stream mode instead of a batch mode
+// all segment ids are validated before calling this function
+func searchSegmentsStreamly(ctx context.Context,
+	mgr *Manager,
+	segments []Segment,
+	segType SegmentType,
+	searchReq *SearchRequest,
+	resultStream chan *SearchResult,
+	errStream chan error,
+) {
+	var (
+		wg sync.WaitGroup
+		// For log only
+		mu                   sync.Mutex
+		segmentsWithoutIndex []int64
+	)
+
+	searchLabel := metrics.SealedSegmentLabel
+	if segType == commonpb.SegmentState_Growing {
+		searchLabel = metrics.GrowingSegmentLabel
+	}
+
+	streamCredits := make(chan int, defaultStreamConcurrency)
+	for i := 0; i < defaultStreamConcurrency; i++ {
+		streamCredits <- i
+	}
+	defer close(streamCredits)
+	searcher := func(seg Segment) error {
+		// record search time
+		tr := timerecord.NewTimeRecorder("searchOnSegments")
+		searchResult, err := seg.Search(ctx, searchReq)
+		errStream <- err
+		resultStream <- searchResult
+		// update metrics
+		elapsed := tr.ElapseSpan().Milliseconds()
+		metrics.QueryNodeSQSegmentLatency.WithLabelValues(fmt.Sprint(paramtable.GetNodeID()),
+			metrics.SearchLabel, searchLabel).Observe(float64(elapsed))
+		metrics.QueryNodeSegmentSearchLatencyPerVector.WithLabelValues(fmt.Sprint(paramtable.GetNodeID()),
+			metrics.SearchLabel, searchLabel).Observe(float64(elapsed) / float64(searchReq.getNumOfQuery()))
+		return nil
+	}
+
```

Comment:
Serially at a certain concurrency in the search task and stream reduce these search results gradually.
The concurrency is defined by streamCredits which is a channel len=defaultStreamConcurrency, in the following pr, this concurrency will be aligned with the memory avaiable of LRU Cache